services/scabbard/libscabbard/src/service/v3/timer_handler.rs

```diff
@@ -12,33 +12,76 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use log::info;
-
 use splinter::{
     error::InternalError,
     service::{FullyQualifiedServiceId, MessageSender, TimerHandler},
+    store::command::StoreCommandExecutor,
 };
 
 use crate::protocol::v3::message::ScabbardMessage;
+use crate::store::{
+    alarm::AlarmType, context::ConsensusContext, event::ConsensusEvent, two_phase_commit::Event,
+    ScabbardStore,
+};
 
-#[derive(Default)]
-pub struct ScabbardTimerHandler {}
+use super::ConsensusRunner;
 
-impl ScabbardTimerHandler {
-    pub fn new() -> Self {
-        Self {}
+pub struct ScabbardTimerHandler<E>
+where
+    E: StoreCommandExecutor + 'static,
+{
+    consensus_runner: ConsensusRunner<E>,
+    store: Box<dyn ScabbardStore>,
+}
+
+impl<E: StoreCommandExecutor + 'static> ScabbardTimerHandler<E> {
+    pub fn new(consensus_runner: ConsensusRunner<E>, store: Box<dyn ScabbardStore>) -> Self {
+        Self {
+            consensus_runner,
+            store,
+        }
     }
 }
 
-impl TimerHandler for ScabbardTimerHandler {
+impl<E: StoreCommandExecutor + 'static> TimerHandler for ScabbardTimerHandler<E> {
     type Message = ScabbardMessage;
 
     fn handle_timer(
         &mut self,
         _sender: &dyn MessageSender<Self::Message>,
         service: FullyQualifiedServiceId,
     ) -> Result<(), InternalError> {
-        info!("handling scabbard timer for: {}", service);
-        Ok(())
+        // Check store if we have an alarm that have expired
+        if let Some(alarm) = self
+            .store
+            .get_alarm(&service, &AlarmType::TwoPhaseCommit)
+            .map_err(|err| InternalError::from_source(Box::new(err)))?
+        {
+            // If the alarm has passed, elapsed will return OK and a duation for how long ago
+            // it was. If it is still in the fututr an error is returned with how long until
+            // the alarm has passed. Therefore if ok, the alarm has passed and an alarm event
+            // should be added to state.
+            if alarm.elapsed().is_ok() {
+                let alarm_event = ConsensusEvent::TwoPhaseCommit(Event::Alarm());
+                let epoch = match self
+                    .store
+                    .get_current_consensus_context(&service)
+                    .map_err(|err| InternalError::from_source(Box::new(err)))?
+                {
+                    Some(ConsensusContext::TwoPhaseCommit(context)) => *context.epoch(),
+                    _ => {
+                        return Err(InternalError::with_message(
```

Comment:
Epoch shouldn't be a factor here, and this runtime error shouldn't be required.

Reply:
This is a larger discussion, as the epoch is required for `store.add_consensus_event`.

Reply:
A story was added to resolve this

Reply:
that story was complete, so this has been removed